gdx/src/com/badlogic/gdx/math/Octree.java

```diff
@@ -0,0 +1,382 @@
+	 *
+	 * @param boxes
+	 */
+	public ObjectSet<BoundingBox> getNodesBoxes (ObjectSet<BoundingBox> boxes) {
+		root.getBoundingBox(boxes);
+		return boxes;
+	}
+
+	protected class OctreeNode {
+
+		private int level;
+
+		// Represents the boundary of the node
+		private final BoundingBox bounds;
+		private final Array<OctreeNode> children = new Array<>(8);
+		private final Array<T> geometries = new Array<T>(maxItemsPerNode);
+
+		public OctreeNode () {
+			bounds = new BoundingBox();
+		}
+
+		private void split () {
+			float midx = (bounds.max.x + bounds.min.x) * 0.5f;
+			float midy = (bounds.max.y + bounds.min.y) * 0.5f;
+			float midz = (bounds.max.z + bounds.min.z) * 0.5f;
+
+			int deeperLevel = level - 1;
+
+			children.add(createNode(new Vector3(bounds.min.x, midy, midz), new Vector3(midx, bounds.max.y, bounds.max.z), deeperLevel));
+			children.add(createNode(new Vector3(midx, midy, midz), new Vector3(bounds.max.x, bounds.max.y, bounds.max.z), deeperLevel));
+			children.add(createNode(new Vector3(midx, midy, bounds.min.z), new Vector3(bounds.max.x, bounds.max.y, midz), deeperLevel));
+			children.add(createNode(new Vector3(bounds.min.x, midy, bounds.min.z), new Vector3(midx, bounds.max.y, midz), deeperLevel));
+			children.add(createNode(new Vector3(bounds.min.x, bounds.min.y, midz), new Vector3(midx, midy, bounds.max.z), deeperLevel));
+			children.add(createNode(new Vector3(midx, bounds.min.y, midz), new Vector3(bounds.max.x, midy, bounds.max.z), deeperLevel));
+			children.add(createNode(new Vector3(midx, bounds.min.y, bounds.min.z), new Vector3(bounds.max.x, midy, midz), deeperLevel));
+			children.add(createNode(new Vector3(bounds.min.x, bounds.min.y, bounds.min.z), new Vector3(midx, midy, midz), deeperLevel));
+
+			// Move geometries from parent to children
+			for (OctreeNode child : children) {
+				for (T geometry : this.geometries) {
+					child.add(geometry);
+				}
+			}
+			this.geometries.clear();
+		}
+
+		private void merge() {
+			for (OctreeNode node : children) {
+				geometries.addAll(node.geometries);
+				freeNode(node);
+			}
+			children.clear();
+		}
+
+		private OctreeNode createNode(Vector3 min, Vector3 max, int level) {
+			OctreeNode node = nodePool.obtain();
+			node.bounds.set(min, max);
+			node.level = level;
+			return node;
+		}   
+
+		private void freeNode(OctreeNode node) {
+			node.geometries.clear();
+			node.children.clear();
+			nodePool.free(node);
+		}
+
+		protected boolean contains (Collider<T> boundingBoxHandler, T geometry) {
+			return boundingBoxHandler.intersects(bounds, geometry);
+		}
+
+		protected void add (T geometry) {
+			if (!collider.intersects(bounds, geometry)) {
+				return;
+			}
+
+			// If is not leaf, check children
+			if (!isLeaf()) {
+				for (OctreeNode child : children) {
+					child.add(geometry);
+				}
+			} else {
+				geometries.add(geometry);
+				if (geometries.size > maxItemsPerNode && level > 0) {
```

Comment:
better check size before adding new item to avoid unnecessary array resizing.

Reply:
@mgsx-dev 
About that, may not sound optimal, but adding the geometry first, if there is a split, the geometry is added to the leaf nodes inside the split method. If I change the order, the geometry needs to be added to the children. Do you think it's better?

```
if (geometries.size > maxItemsPerNode && level > 0) {
    split();
    for (OctreeNode child : children) {
        child.add(geometry);
    }
} else {
    geometries.add(geometry);
}
```

Reply:
`if (geometries.size >= maxItemsPerNode && level > 0)` :-)
yeah, that's fine, it's optimal. The problem was that we initialize geometries with 
maxItemsPerNode capacity, so adding new item before required to resize and copy the whole array for nothing.

Reply:
Oh, I missed that, you are right!
This change is hard to decide, I am more inclined to leave it as is.

Reply:
the code you proposed in your comment does fix it except it should use >= check instead of >

Reply:
Done!

Reply:
i think you forgot the else from your snippet